### PR TITLE
feat: add ClientStoreProvider interface for isolated prefix stores

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

Adds a nice slim interface with a single method `ClientStore` for light client modules to retrieve a prefixed client store.

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/apps/callbacks/testing/simapp/app.go [hunk 1]
```diff
@@ -569,11 +569,11 @@ func NewSimApp(
 
 	clientRouter := app.IBCKeeper.ClientKeeper.GetRouter()
 
-	tmLightClientModule := ibctm.NewLightClientModule(appCodec, keys[ibcexported.StoreKey], authtypes.NewModuleAddress(govtypes.ModuleName).String())
```
Comment (colin-axner) [inline modules/apps/callbacks/testing/simapp/app.go:572]: I like this wiring, very simple

File: modules/core/02-client/keeper/keeper.go [hunk 2]
```diff
@@ -38,11 +38,11 @@ type Keeper struct {
 }
 
 // NewKeeper creates a new NewKeeper instance
-func NewKeeper(cdc codec.BinaryCodec, key storetypes.StoreKey, router *types.Router, legacySubspace types.ParamSubspace, sk types.StakingKeeper, uk types.UpgradeKeeper) Keeper {
+func NewKeeper(cdc codec.BinaryCodec, key storetypes.StoreKey, legacySubspace types.ParamSubspace, sk types.StakingKeeper, uk types.UpgradeKeeper) Keeper {
```
Comment (colin-axner) [inline modules/core/02-client/keeper/keeper.go:41]: niice, I like not having to pass the router in

File: modules/light-clients/06-solomachine/internal/keeper/keeper.go [hunk 3]
```diff
@@ -1,38 +1,23 @@
...
-	storetypes "cosmossdk.io/store/types"
-
 	"github.com/cosmos/cosmos-sdk/codec"
-	sdk "github.com/cosmos/cosmos-sdk/types"
-
-	host "github.com/cosmos/ibc-go/v8/modules/core/24-host"
 )
 
 // Keeper defines the 06-solomachine Keeper.
+// TODO(damian): delete the keeper and put the codec on the LightClientModule?
```
Comment (colin-axner) [inline modules/light-clients/06-solomachine/internal/keeper/keeper.go:8]: +1 keeper may only be necessary if a msg server is added (which for solo machine I'm not sure we need)

File: modules/light-clients/09-localhost/light_client_module.go [hunk 4]
```diff
@@ -12,16 +11,23 @@ import (
 	"github.com/cosmos/ibc-go/v8/modules/light-clients/09-localhost/internal/keeper"
 )
 
+var _ exported.LightClientModule = (*LightClientModule)(nil)
+
 type LightClientModule struct {
-	keeper keeper.Keeper
+	keeper        keeper.Keeper
+	storeProvider exported.ClientStoreProvider
```
Comment (colin-axner) [inline modules/light-clients/09-localhost/light_client_module.go:18]: localhost we should actually just provide the full ibc store. I don't think it needs a store provider

**Review by colin-axner — APPROVED** [review]
Excellent!! :100: I like this approach a lot better than passing our the ibc store key. I think it aligns well with the pr I opened to remove genesis handling from light client modules. I think the general narrative is that light client modules manage their client type prefix in the client store, but that they must request access to these stores from 02-client. 02-client in exchange will manage genesis import/export

I'm happy with the API naming, we can always revisit if someone prefers a different name

File: modules/core/02-client/keeper/keeper.go (hunk 2 above)
Comment (damiannolan) [reply modules/core/02-client/keeper/keeper.go:41]: aye!! there's possibly more improvements to be made on structuring and what not but let's go with this for now

Comment (damiannolan) [reply modules/core/02-client/keeper/keeper.go:41]: Could also instantiate the storeProvider here when creating the router.

**Pull request merged**